// eslint-plugin-office-addins/lib/rules/no-context-sync-in-loop.js
module.exports = {
    name: "no-context-sync-in-loop",
    meta: {
        type: "problem",
        messages: {
            loopedSync: "Calling context.sync() inside a loop can lead to bad performance",
        },
        docs: {
            description: "Calling context.sync() inside of a loop dramatically increates the time the code runs the more iterations that are run",
            category: "Best Practices",
            recommended: false,
            url: "https://docs.microsoft.com/en-us/office/dev/add-ins/concepts/correlated-objects-pattern?view=powerpoint-js-1.1",
        },
        schema: [],
    },
    create: function (context) {
        return {
            "CallExpression[callee.property.name='run'] :matches(ForStatement, ForInStatement, WhileStatement, DoWhileStatement, ForOfStatement) CallExpression[callee.object.name='context'][callee.property.name='sync']"(node) {
                context.report({
                    node: node.callee,
                    messageId: "loopedSync",
                });
            },
        };
    },
};
